### DISCUSSION TITLE
Detach switch, binary sensor state

### DISCUSSION THREAD

**Discussion by Brkie** [arendst/Tasmota#16693] [Q&A]
Hi,

I have a tasmota that i have detached, so now I have a switch.entity and a binary_sensor.entity in Home Assistant. When I use the psyhical (real life) switch, the binary sensor changes state to on and off. All good. I then created an automation that when binary_sensor goes on -> turn on Zigbee light. Binary sensor Off > turn off zigbee light.

But if I turn on my zigbee light thru homekit or google home, the binary sensor stays off. 

To my problem - when I want to use the psyhical switch - i first have to toggle it, so the binary sensor goes on, then toggle again so it gets off.

Is there any way to fix this - like make an automation that binary sensor mirrors the state of my Zigbee light or should I do this some other way...? 

Thx

**Comment by barbudor** [comment]
A Tasmota Switch reports it's input.
As you cannot flip the mechanical switch when you change your light status, Tasmota cannot change it's state.

However you may be able to configure the switch to report a TOGGLE everytime it is flipped instead of reporting an ON or OFF state
This is done by using the command `SwitchMode 0`

**Reply by Brkie** [reply]
Ok, I now have a toggle switch. But now neather the switch or the binary sensors toggles..?

**Reply by Brkie** [reply]
After a restart of tasmota, Binary sensor is removed. I used `SwitchMode 15` before.

**Reply by barbudor** [reply]
You didn't say that before
May be use SetOption114 to detach the switch along SwitchMode0 ?

I don't know how HA works but what you are asking is in my understanding exclusive of a binary sensor
A binary sensor has 2 positions : 0/1, On/Off
As such it reflects the state of its input

A switch in SwitchMode0 will inform HA of the flipping of the switch independently of the actual position of the switch (which you actually don't care).
So that means that your HA configuration should be different to react to TOGGLE instead of ON/OFF

**Reply by Brkie** [reply]
Sorry, I tought that u knew that I use SwitchMode 15 cuz I hade a binary sensor. I just found something else that can help me on an easyer way.

Can I make input state toggle on every press and send a MQTT topic like "zigbee2mqtt/MyZigbeeLight/set" with payload "TOGGLE"

Can you maybe point me in that direction? That should solve my problem and then I even dont need the binary sensor lol

**Reply by barbudor** [reply]
There are different ways of detaching switches : rules, switchmode1 and Setoption114. You didn't specified which one. And "binary sensor" is a HA notion, not Tasmota.

Is your mechanical switch a rocker switch that flips between On and Off, and you want to send the Zigbee toggle everytime the switch is flipped to the other side ?
Or is it a momentary push-button that you press and release and you want the Zigbee togle mesage to be sent when you press it ?

In case 1), you should use SwitchMode0 so that each flip makes a TOGGLE event
You can capture the event with a rule like:
```
Rule1 on switch1#state=2 do publish zigbee2mqtt/MyZigbeeLight/set TOGGLE endon
Rule1 1
```

In case 2, you need to use `SwitchMode 3`or `SwitchMode 4`. Depending on how you switch is connected, the TOGGLE event will occur either when you press or when you release the push button.
The same rule as above will capture the TOGGLE event and send you MQTT message.

**Reply by Brkie** [reply]
Thats what I was looking for. Its case 1. Thx alot. I just have one more problem and then i think i got it. I use a space in a topic name (dumb, I know lol) - the console makes my rule use _ instaed of space. 

So the topic is: `zigbee2mqtt/Mileas taklampa/set`
the console gives me this:
`21:44:13.953 CMD: Rule1 on switch1#state=2 do publish zigbee2mqtt/Mileas taklampa/set TOGGLE endon
21:44:13.959 RUL: Stored uncompressed, would compress from 74 to 55 (-26%)
21:44:13.964 MQT: stat/tasmota_FCF1AC/RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":74,"Free":437,"Rules":"on switch1#state=2 do publish zigbee2mqtt/Mileas taklampa/set TOGGLE endon"}}
`

this is result when I toggle:
`21:44:52.846 MQT: cmnd/zigbee2mqtt/Mileas_taklampa/set/POWER1 = TOGGLE`

(btw where did `POWER1`come from lol)

I read somewhere that u could use # as a separator but that didnt help. Man thx alot, lol I was so lost in the beggining but now I start to see a light in the tunel öol

**Reply by barbudor** [reply]
I have no idea why you get that POWER and that _

At least I would say that using a space in the topic is bad because the publish command should see the space as a separation between the topic and the payload.

A basic publish test in the console gives me:
```
23:10:45.509 CMD: publish zzz/abc def/xyz TOGGLE
23:10:45.515 MQT: zzz/abc = def/xyz TOGGLE
```
Adding quotes doesn't help:
```
23:11:43.816 CMD: publish "zzz/abc def/xyz" TOGGLE
23:11:43.821 MQT: "zzz/abc = def/xyz" TOGGLE
```
And in a rule makes no difference:
```
Rule3 on event#test do publish zzz/abc def/xyz TOGGLE endon
Rule3 1
event test
23:12:41.058 CMD: event test
23:12:41.067 MQT: stat/skt-barbuxps/EVENT = {"Event":"Done"}
23:12:41.168 RUL: EVENT#TEST performs "publish zzz/abc def/xyz TOGGLE"
23:12:41.174 MQT: zzz/abc = def/xyz TOGGLE
```

`#` is a separator in rules triggers between levels like the dot `.` in some programming language
switch1#state is somehow the 'state' of object 'switch1'

When using rules to triggers on a SENSOR message like `{ "BME280": { "Temperature": 23.7, "Humidity": 56.1 } }`
You would use `ON BME280#Temperature>27 do POWER ON endon` to start the fan above 27°
